Array additions

glennsl opened the pull request:
This adds:

- `make`, seems necessary to be able to create initialized arrays without imperatively pushing onto an empty array.
- `init`, is `make` with an initializer function instead of a constant.
- `findMap`, is surprisingly useful, to find an element in a nested array, for example (`[[1, 2], [3]]->findMap(find(_, el => ...))`), or to find the first element in an `array<option<_>>` that is not `None`.
- `keepSome`, unwraps the options in an `array<option<_>>` while discarding the `None`s.

It also replaces the reimplementation of flatMap with a binding to `Array.prototype.flatMap`.

This is, to begin with, just to gather some feedback on the relevance, naming and implementations, before I add documentation and tests.

TODO:
- [x] Documentation
- [x] Tests
- [x] Changelog

cknitt commented on src/Core__Array.res:
```diff
@@ -11,6 +13,28 @@ external fromArrayLikeWithMap: (Js.Array2.array_like<'a>, 'a => 'b) => array<'a>
+    []
+  } else {
+    let arr = makeUninitializedUnsafe(len)
+    for i in 0 to len - 1 {
```
I would propose to do `arr->fillAllInPlace(x)` instead of the loop.

cknitt commented on src/Core__Array.res:
```diff
@@ -11,6 +13,28 @@ external fromArrayLikeWithMap: (Js.Array2.array_like<'a>, 'a => 'b) => array<'a>
 @val external fromIterator: Core__Iterator.t<'a> => array<'a> = "Array.from"
 @val external fromIteratorWithMap: (Core__Iterator.t<'a>, 'a => 'c) => array<'a> = "Array.from"
 
+let make = (len, x) =>
```
Maybe a named argument `~length` should be used here.

zth replied:
I agree, I think that might make it clearer.

glennsl replied:
Should this apply to `init` too then? Or is it just because both arguments could be of the same type in this case?

cknitt replied:
I think it should apply to both functions.

glennsl replied:
Added in https://github.com/rescript-association/rescript-core/pull/49/commits/cef5768e050ef45d4b2f3e31087d1db94ae11ebb

cknitt commented on src/Core__Array.res:
```diff
@@ -11,6 +13,28 @@ external fromArrayLikeWithMap: (Js.Array2.array_like<'a>, 'a => 'b) => array<'a>
+    arr
+  }
+
+let init = (len, f) =>
```
As this creates a new array, too, the function name should start with `make` IMHO, maybe `makeWithInitializer` or something like that.

zth replied:
Agree, for the consistency.

glennsl replied:
Or perhaps `fromInitializer`?

cknitt replied:
`fromInitializer` sounds good to me, too.

zth replied:
✅ `fromInitializer` it is.

glennsl replied:
Changed in https://github.com/rescript-association/rescript-core/pull/49/commits/889e993f80fd765776fd414cec6bcd69eb4fb4b0

cknitt commented on src/Core__Array.res:
```diff
@@ -194,7 +216,22 @@ let filterMapU = (a, f) => {
+
+@send external flatMap: (array<'a>, 'a => array<'b>) => array<'b> = "flatMap"
+
+let findMap = (arr, f) => {
```
There was a discussion in #4 where we decided not to add this function for now.

zth replied:
I'm open to revisiting that decision though. I still think `findMap` is useful and along the lines of the type of "extra" functionality we could support in `Core`, so if more agree then I don't see any harm in adding it.

glennsl replied:
Not sure what the threshold is, but I've mentioned two use cases here that weren't mentioned in that issue at least. I think this makes more sense in rescript than in JS because we're dealing explicitly with `option`s, whereas making `null` values significant in JS doesn't work since they don't compose (nest) properly.

zth replied:
Yeah I think adding it won't hurt, given that it's about handling options, and we're trying to make that as smooth as we can. Ok with you @cknitt ?

cknitt replied:
👍

cknitt reviewed (commented):
Thanks! Here is some feedback from my part.

What do you think @zth?

zth commented on src/Core__Array.resi:
```diff
@@ -100,10 +102,11 @@ external setUnsafe: (array<'a>, int, 'a) => unit = "%array_unsafe_set"
 let shuffleInPlace: array<'a> => unit
-let flatMap: (array<'a>, 'a => array<'b>) => array<'b>
-
+@send external flatMap: (array<'a>, 'a => array<'b>) => array<'b> = "flatMap"
```
Thanks for binding to the real thing here instead. IIRC I opted to use the last variant because I believe that's what Belt used, but I think the "real" `flatMap` is widely supported now so this should be good.

zth commented on src/Core__Array.resi:
```diff
@@ -100,10 +102,11 @@ external setUnsafe: (array<'a>, int, 'a) => unit = "%array_unsafe_set"
 let findIndexOpt: (array<'a>, 'a => bool) => option<int>
 let reverse: array<'a> => array<'a>
 let filterMap: (array<'a>, 'a => option<'b>) => array<'b>
+let keepSome: array<option<'a>> => array<'a>
```
I like this. The naming also makes sense. But, just to make sure we're being as consistent as we can - would there be some way to name this using `filter` instead of `keep`, that would make as much sense as `keepSome`? It'd be for consistency, we for example changed the name of Belt's `keepMap` to `filterMap` just to be as close to the JS naming convention as we can.

`filterSome` feels like it implies the opposite (remove all `Some`). `filterNone` also not 100%. What do you think?

cknitt replied:
Good point. I am would be fine with `filterSome`.

glennsl replied:
You describe my thoughts pretty much exactly @zth. While I found the `keep` naming scheme unintuitive in `Belt`, I think it describes it pretty well here. `filterNone` would be my runner-up, as `filterSome` seems to me to imply the opposite.

More generally, I think a desire for consistency can be a bit of a trap. I've seen so many situations where people have made themselves such nice mental models that when reality contradicts it they try to force the real world to conform to their theories. The entire field of Economics comes to mind as one example of this :laughing: It's good to keep in mind that consistency isn't the goal, but a means to an end, and stops working towards that end when it becomes unintuitive. But it's also a good indicator to take pause and not rush the decision. I can just split it out into its own PR if necessary to merge the rest.

glennsl replied:
Elm has it as [`Maybe.Extra.values`](https://package.elm-lang.org/packages/elm-community/maybe-extra/latest/Maybe-Extra#values), btw. I think, as a rule, functions ought to live in the module of the more general type they operate on though. If for nothing else, it would be awkward to have functions operating on both `array<option<_>>` and `list<option<_>>` in the same module with a same naming scheme I think.

glennsl replied:
It might be a good idea to move controversial or experimental functions like this to an incubator project, such as `rescript-core-extras`, for example, to provide different levels of stability guarantees.

zth replied:
Oh yeah I agree about consistency and naming, not trying to be consistent just to be consistent if it produces a less understandable result. But it's with thinking a round or two on the naming now that we have a fairly set naming scheme.

@cknitt what do you think about filterNone vs keepSome?

As for the function itself, I think this is a great addition. I know the pain all too well from working in TS occasionally, where there's no built in way to clean an array of nullables that TS understand, so you need to roll your own assertion function that "tricks" the type system.

cknitt replied:
I also think the function is useful as a shortcut for `array->Array.filterMap(x => x)`. I like `keepSome` better than `filterNone`.

zth replied:
That settles it then. Let's keep the original naming proposed 👍

zth commented:
Thanks @glennsl, these additions look quite useful. I added my comments to the discussion as well.

glennsl commented:
I suppose this could be merged now, since neither documentation for the rest of the module or a test framework is in place yet. It could also be what initiates both. Up to you!

zth commented:
If you don't mind, I think adding both in this PR would be good, so it doesn't get left behind. Would that be OK? Let me know how I can help out if needed.

zth commented:
Also, a change log entry would be good as well.

glennsl commented:
Sure. Do you have any thoughts on testing framework? Should I pull in something like `rescript-jest` or make something very simple and bespoke?

zth commented:
We have a simple one already in `Test.res` that we can start by using. Check out `PromiseTest.res` for how to use it.

Thank you for all your contributions so far by the way, very valuable! 👍

glennsl commented:
Added documentation,test and updated changelog. Also fixed `reduce` and` reduceReverse`, which were reimplemented instead of binding to its native counterparts. That did require changing the type signatures and renaming `reduceReverse` to `reduceRight` though.

zth commented:
Great! Taking a look again soon.

zth commented on src/Core__Array.resi:
````diff
@@ -101,11 +141,34 @@ external setUnsafe: (array<'a>, int, 'a) => unit = "%array_unsafe_set"
 let reverse: array<'a> => array<'a>
 let filterMap: (array<'a>, 'a => option<'b>) => array<'b>
+
+/**
+  `fromInitializer(~length, f)`
+
+  Creates an array of length `length` initialized with the value returned from `f ` for each index.
+
+  ```res example
+  Array.make(~length=3, i => i + 3) == [3, 4, 5]
+  ```
+*/
+let keepSome: array<option<'a>> => array<'a>
````
This doesn't seem to be the right docstring?

glennsl replied:
I actually just forgot to save the file after the last edit :facepalm: Fixed in https://github.com/rescript-association/rescript-core/pull/49/commits/b80a80f10b4c925d6f91fe06c7493a876efc64a8

zth reviewed (commented):
Actually, `reduce` and `reduceReverse` were intentionally taken from `Belt` instead of using their native counterparts because the `Belt` version allows for full type inference whereas the native one does not. So that was a deliberate decision.

Just had a small comment, other than that and the reduce stuff this is good to go from my end. Thanks!

cc @cknitt if you want to have a last look as well.

cknitt commented:
> Actually, `reduce` and `reduceReverse` were intentionally taken from `Belt` instead of using their native counterparts because the `Belt` version allows for full type inference whereas the native one does not. So that was a deliberate decision.

If it's just for the improved argument order for better type inference, then we could simply do

```rescript
@send external reduce: (array<'b>, ('a, 'b) => 'a, 'a) => 'a = "reduce"

let reduce = (arr, x, f) => reduce(arr, f, x)
```

(same for `reduceRight`).

glennsl commented:
Indeed. Seems odd to have them reimplemented and with different names when only a reordering of the arguments is needed.

Could probably also just label the argument to allow it to be moved before the callback function?

```
@send external reduce: (array<'b>, ('a, 'b) => 'a, ~init: 'a) => 'a = "reduce"
```

glennsl commented:
So, reorder, not label?

zth commented:
Correct.

glennsl commented:
Arguments reordered! Also reordered the type variable names for consistency.

zth commented:
Fantastic, great work @glennsl ! Good to go from my end. @cknitt ?

cknitt commented:
Yes, great work! I think it would be good to use pipe `->` in the examples, but that can also be done separately later as it applies to other modules, too. Let's merge now! 🙂
